Juju 4 beta 7 - cannot deploy Kubeflow on AKS using terraform - credentials not found

### Description

> [!NOTE]
> Preliminary testing for Juju 4 by Solutions QA

1. Deployed Solutions QA's AKS cluster using Terraform (we don't use juju for it)
2. `add-k8s` with the cluster's credentials, and create a new controller for it.
3. Attempt to deploy Charmed Kubeflow+MLFlow using their [terraform solution](https://github.com/canonical/charmed-kubeflow-solutions/tree/track/1.10/modules/kubeflow-mlflow) --- **FAILS**

```console
module.kubeflow.juju_model.kubeflow[0]: Creating...
╷
│ Error: Client Error
│ 
│   with module.kubeflow.juju_model.kubeflow[0],
│   on ../kubeflow/main.tf line 1, in resource "juju_model" "kubeflow":
│    1: resource "juju_model" "kubeflow" {
│ 
│ Unable to create model "\"kubeflow\"", got error: cloud credential for new
│ model is not valid (not found)
╵
Error: Process completed with exit code 1.
```

### Juju version

4 beta 7

### Cloud

Kubernetes

### Expected behaviour

The Kubeflow+MLFlow deployment using Terraform works with Juju 4/beta as it already does with Juju 3/stable.

### Reproduce / Test

See above, also reproducible in SQA tests.

### Notes & References

The specific run in SQA: https://solutions.qa.canonical.com/testruns/78b8c172-5646-46f4-9ca2-4fb9dcab0696

Also, I am not sure if the cloud should be Kubernetes in this case or Azure.  AKS is deployed by SQA without using Juju, and then it is `add-k8s`-ed.

--- a/apiserver/facades/client/modelmanager/modelmanager.go
+++ b/apiserver/facades/client/modelmanager/modelmanager.go
@@ -23,10 +23,11 @@ import (
 	"github.com/juju/juju/core/permission"
 	"github.com/juju/juju/core/semversion"
 	corestatus "github.com/juju/juju/core/status"
-	"github.com/juju/juju/core/user"
+	coreuser "github.com/juju/juju/core/user"
 	"github.com/juju/juju/domain/access"
 	accesserrors "github.com/juju/juju/domain/access/errors"
 	clouderrors "github.com/juju/juju/domain/cloud/errors"
+	credentialerrors "github.com/juju/juju/domain/credential/errors"
 	"github.com/juju/juju/domain/model"
 	modelerrors "github.com/juju/juju/domain/model/errors"
 	"github.com/juju/juju/environs/config"
@@ -220,9 +221,21 @@ func (m *ModelManagerAPI) CreateModel(ctx context.Context, args params.ModelCrea
 			return result, errors.Trace(err)
 		}
 		creationArgs.Credential = credential.KeyFromTag(cloudCredentialTag)
+	} else {
+		// we're looking up credentials for the apiUser - what if JIMM is creating a model on behalf of another user (JIMM always
+		// specifies the credential, so we should be fine)
+		credKey, err := m.modelService.DefaultCloudCredentialKeyForOwner(ctx, coreuser.NameFromTag(m.apiUser), cloudTag.Id())
+		if err != nil {
+			// if we cannot find the default credential key, we continue as some clouds may allow model creation without
+			// credentials, so we error out only if this is NOT a NotFound error
+			if !errors.Is(err, credentialerrors.NotFound) {
+				return result, errors.Trace(err)
+			}
+		}
+		creationArgs.Credential = credKey
 	}
 
-	userUUID, err := m.accessService.GetUserUUIDByName(ctx, user.NameFromTag(m.apiUser))
+	userUUID, err := m.accessService.GetUserUUIDByName(ctx, coreuser.NameFromTag(m.apiUser))
 	switch {
 	case errors.Is(err, accesserrors.UserNotFound):
 		return result, internalerrors.Errorf(
@@ -238,7 +251,7 @@ func (m *ModelManagerAPI) CreateModel(ctx context.Context, args params.ModelCrea
 			m.apiUser.Name(), args.Name, err,
 		)
 	}
-	creationArgs.AdminUsers = []user.UUID{userUUID}
+	creationArgs.AdminUsers = []coreuser.UUID{userUUID}
 
 	// Create the model in the controller database.
 	modelUUID, activator, err := m.modelService.CreateModel(ctx, creationArgs)
@@ -533,7 +546,7 @@ func (m *ModelManagerAPI) listModelSummariesForUser(ctx context.Context, tag nam
 		)
 	}
 	result := params.ModelSummaryResults{}
-	userUUID, err := m.accessService.GetUserUUIDByName(ctx, user.NameFromTag(tag))
+	userUUID, err := m.accessService.GetUserUUIDByName(ctx, coreuser.NameFromTag(tag))
 	if err != nil {
 		return result, makeErrorReturn(err)
 	}
@@ -689,7 +702,7 @@ func (m *ModelManagerAPI) ListModels(ctx context.Context, userEntity params.Enti
 		return result, errors.Trace(err)
 	}
 
-	userUUID, err := m.accessService.GetUserUUIDByName(ctx, user.NameFromTag(userTag))
+	userUUID, err := m.accessService.GetUserUUIDByName(ctx, coreuser.NameFromTag(userTag))
 	if err != nil {
 		return result, errors.Trace(err)
 	}
@@ -709,7 +722,7 @@ func (m *ModelManagerAPI) ListModels(ctx context.Context, userEntity params.Enti
 
 	for _, mi := range models {
 		var lastConnection *time.Time
-		lc, err := m.accessService.LastModelLogin(ctx, user.NameFromTag(userTag), mi.UUID)
+		lc, err := m.accessService.LastModelLogin(ctx, coreuser.NameFromTag(userTag), mi.UUID)
 		if errors.Is(err, accesserrors.UserNeverAccessedModel) {
 			lastConnection = nil
 		} else if errors.Is(err, modelerrors.NotFound) {
@@ -963,7 +976,7 @@ func (m *ModelManagerAPI) getModelInfo(ctx context.Context, modelUUID coremodel.
 		}
 	}
 
-	info.Users, err = commonmodel.ModelUserInfo(ctx, m.modelService, modelTag, user.NameFromTag(m.apiUser), m.isAdmin)
+	info.Users, err = commonmodel.ModelUserInfo(ctx, m.modelService, modelTag, coreuser.NameFromTag(m.apiUser), m.isAdmin)
 	if err != nil {
 		return params.ModelInfo{}, errors.Annotate(err, "getting model user info")
 	}
@@ -1036,7 +1049,7 @@ func (m *ModelManagerAPI) ModifyModelAccess(ctx context.Context, args params.Mod
 				Access: modelAccess,
 			},
 			Change:  permission.AccessChange(arg.Action),
-			Subject: user.NameFromTag(targetUserTag),
+			Subject: coreuser.NameFromTag(targetUserTag),
 		})
 
 		result.Results[i].Error = apiservererrors.ServerError(err)

--- a/apiserver/facades/client/modelmanager/modelmanager_test.go
+++ b/apiserver/facades/client/modelmanager/modelmanager_test.go
@@ -28,7 +28,7 @@ import (
 	"github.com/juju/juju/core/permission"
 	"github.com/juju/juju/core/semversion"
 	corestatus "github.com/juju/juju/core/status"
-	"github.com/juju/juju/core/user"
+	coreuser "github.com/juju/juju/core/user"
 	usertesting "github.com/juju/juju/core/user/testing"
 	jujuversion "github.com/juju/juju/core/version"
 	"github.com/juju/juju/domain/access"
@@ -181,7 +181,7 @@ func (s *modelManagerSuite) expectCreateModel(
 
 	defaultCred := credential.Key{
 		Cloud: "dummy",
-		Owner: user.AdminUserName,
+		Owner: coreuser.AdminUserName,
 		Name:  "some-credential",
 	}
 
@@ -197,7 +197,7 @@ func (s *modelManagerSuite) expectCreateModel(
 	s.modelService.EXPECT().CreateModel(gomock.Any(), domainmodel.GlobalModelCreationArgs{
 		Name:        modelCreateArgs.Name,
 		Qualifier:   "admin",
-		AdminUsers:  []user.UUID{adminUUID},
+		AdminUsers:  []coreuser.UUID{adminUUID},
 		Cloud:       expectedCloudName,
 		CloudRegion: expectedCloudRegion,
 		Credential:  expectedCloudCredential,
@@ -312,7 +312,7 @@ func (s *modelManagerSuite) TestCreateModelArgsWithCloud(c *tc.C) {
 
 	cloudCredental := credential.Key{
 		Cloud: "dummy",
-		Owner: user.AdminUserName,
+		Owner: coreuser.AdminUserName,
 		Name:  "some-credential",
 	}
 	args := params.ModelCreateArgs{
@@ -342,8 +342,9 @@ func (s *modelManagerSuite) TestCreateModelDefaultRegion(c *tc.C) {
 		Qualifier: "admin",
 	}
 
-	s.expectCreateModel(c, ctrl, args, credential.Key{}, "dummy", "dummy-region")
+	s.expectCreateModel(c, ctrl, args, credential.Key{Cloud: "dummy", Owner: coreuser.AdminUserName, Name: "some-credential"}, "dummy", "dummy-region")
 	s.modelInfoService.EXPECT().CreateModel(gomock.Any()).Return(nil)
+	s.modelService.EXPECT().DefaultCloudCredentialKeyForOwner(gomock.Any(), usertesting.GenNewName(c, "admin"), "dummy").Return(credential.Key{Name: "some-credential", Cloud: "dummy", Owner: usertesting.GenNewName(c, "admin")}, nil)
 
 	_, err := s.api.CreateModel(c.Context(), args)
 	c.Assert(err, tc.ErrorIsNil)
@@ -358,8 +359,9 @@ func (s *modelManagerSuite) TestCreateModelDefaultCredentialAdmin(c *tc.C) {
 		Qualifier: "admin",
 	}
 
-	s.expectCreateModel(c, ctrl, args, credential.Key{}, "dummy", "dummy-region")
+	s.expectCreateModel(c, ctrl, args, credential.Key{Cloud: "dummy", Owner: coreuser.AdminUserName, Name: "some-credential"}, "dummy", "dummy-region")
 	s.modelInfoService.EXPECT().CreateModel(gomock.Any()).Return(nil)
+	s.modelService.EXPECT().DefaultCloudCredentialKeyForOwner(gomock.Any(), usertesting.GenNewName(c, "admin"), "dummy").Return(credential.Key{Name: "some-credential", Cloud: "dummy", Owner: usertesting.GenNewName(c, "admin")}, nil)
 
 	_, err := s.api.CreateModel(c.Context(), args)
 	c.Assert(err, tc.ErrorIsNil)
@@ -371,7 +373,7 @@ func (s *modelManagerSuite) TestCreateModelArgsWithAgentVersion(c *tc.C) {
 
 	cloudCredental := credential.Key{
 		Cloud: "dummy",
-		Owner: user.AdminUserName,
+		Owner: coreuser.AdminUserName,
 		Name:  "some-credential",
 	}
 	args := params.ModelCreateArgs{
@@ -399,7 +401,7 @@ func (s *modelManagerSuite) TestCreateModelArgsWithAgentVersionAndStream(c *tc.C
 
 	cloudCredental := credential.Key{
 		Cloud: "dummy",
-		Owner: user.AdminUserName,
+		Owner: coreuser.AdminUserName,
 		Name:  "some-credential",
 	}
 	args := params.ModelCreateArgs{
@@ -629,7 +631,7 @@ func (s *modelManagerSuite) TestUpdatedModel(c *tc.C) {
 			Access: permission.WriteAccess,
 		},
 		Change:  permission.Grant,
-		Subject: user.NameFromTag(testUser),
+		Subject: coreuser.NameFromTag(testUser),
 	}
 	as.UpdatePermission(gomock.Any(), updateArgs).Return(nil)
 
@@ -771,7 +773,7 @@ func (s *modelManagerSuite) TestListModelsAdminSelf(c *tc.C) {
 	modelUUIDNotExist := modeltesting.GenModelUUID(c)
 
 	now := time.Now()
-	s.accessService.EXPECT().GetUserUUIDByName(gomock.Any(), user.NameFromTag(userTag)).Return(userUUID, nil)
+	s.accessService.EXPECT().GetUserUUIDByName(gomock.Any(), coreuser.NameFromTag(userTag)).Return(userUUID, nil)
 	s.modelService.EXPECT().ListAllModels(gomock.Any()).Return([]coremodel.Model{
 		{UUID: modelUUID, Qualifier: "prod"},
 		{UUID: modelUUIDNeverAccessed, Qualifier: "prod"},
@@ -779,11 +781,11 @@ func (s *modelManagerSuite) TestListModelsAdminSelf(c *tc.C) {
 	}, nil)
 
 	s.accessService.EXPECT().LastModelLogin(
-		gomock.Any(), user.NameFromTag(userTag), modelUUID).Return(now, nil)
+		gomock.Any(), coreuser.NameFromTag(userTag), modelUUID).Return(now, nil)
 	s.accessService.EXPECT().LastModelLogin(
-		gomock.Any(), user.NameFromTag(userTag), modelUUIDNeverAccessed).Return(time.Time{}, accesserrors.UserNeverAccessedModel)
+		gomock.Any(), coreuser.NameFromTag(userTag), modelUUIDNeverAccessed).Return(time.Time{}, accesserrors.UserNeverAccessedModel)
 	s.accessService.EXPECT().LastModelLogin(
-		gomock.Any(), user.NameFromTag(userTag), modelUUIDNotExist).Return(time.Time{}, modelerrors.NotFound)
+		gomock.Any(), coreuser.NameFromTag(userTag), modelUUIDNotExist).Return(time.Time{}, modelerrors.NotFound)
 
 	results, err := s.api.ListModels(
 		c.Context(),
@@ -820,7 +822,7 @@ func (s *modelManagerSuite) TestListModelsNonAdminSelf(c *tc.C) {
 	modelUUIDNotExist := modeltesting.GenModelUUID(c)
 
 	now := time.Now()
-	s.accessService.EXPECT().GetUserUUIDByName(gomock.Any(), user.NameFromTag(userTag)).Return(userUUID, nil)
+	s.accessService.EXPECT().GetUserUUIDByName(gomock.Any(), coreuser.NameFromTag(userTag)).Return(userUUID, nil)
 	s.modelService.EXPECT().ListModelsForUser(gomock.Any(), userUUID).Return([]coremodel.Model{
 		{UUID: modelUUID, Qualifier: "prod"},
 		{UUID: modelUUIDNeverAccessed, Qualifier: "prod"},
@@ -828,11 +830,11 @@ func (s *modelManagerSuite) TestListModelsNonAdminSelf(c *tc.C) {
 	}, nil)
 
 	s.accessService.EXPECT().LastModelLogin(
-		gomock.Any(), user.NameFromTag(userTag), modelUUID).Return(now, nil)
+		gomock.Any(), coreuser.NameFromTag(userTag), modelUUID).Return(now, nil)
 	s.accessService.EXPECT().LastModelLogin(
-		gomock.Any(), user.NameFromTag(userTag), modelUUIDNeverAccessed).Return(time.Time{}, accesserrors.UserNeverAccessedModel)
+		gomock.Any(), coreuser.NameFromTag(userTag), modelUUIDNeverAccessed).Return(time.Time{}, accesserrors.UserNeverAccessedModel)
 	s.accessService.EXPECT().LastModelLogin(
-		gomock.Any(), user.NameFromTag(userTag), modelUUIDNotExist).Return(time.Time{}, modelerrors.NotFound)
+		gomock.Any(), coreuser.NameFromTag(userTag), modelUUIDNotExist).Return(time.Time{}, modelerrors.NotFound)
 
 	results, err := s.api.ListModels(
 		c.Context(),

--- a/apiserver/facades/client/modelmanager/service_mock_test.go
+++ b/apiserver/facades/client/modelmanager/service_mock_test.go
@@ -360,6 +360,45 @@ func (c *MockModelServiceCreateModelCall) DoAndReturn(f func(context.Context, mo
 	return c
 }
 
+// DefaultCloudCredentialKeyForOwner mocks base method.
+func (m *MockModelService) DefaultCloudCredentialKeyForOwner(arg0 context.Context, arg1 user.Name, arg2 string) (credential.Key, error) {
+	m.ctrl.T.Helper()
+	ret := m.ctrl.Call(m, "DefaultCloudCredentialKeyForOwner", arg0, arg1, arg2)
+	ret0, _ := ret[0].(credential.Key)
+	ret1, _ := ret[1].(error)
+	return ret0, ret1
+}
+
+// DefaultCloudCredentialKeyForOwner indicates an expected call of DefaultCloudCredentialKeyForOwner.
+func (mr *MockModelServiceMockRecorder) DefaultCloudCredentialKeyForOwner(arg0, arg1, arg2 any) *MockModelServiceDefaultCloudCredentialKeyForOwnerCall {
+	mr.mock.ctrl.T.Helper()
+	call := mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "DefaultCloudCredentialKeyForOwner", reflect.TypeOf((*MockModelService)(nil).DefaultCloudCredentialKeyForOwner), arg0, arg1, arg2)
+	return &MockModelServiceDefaultCloudCredentialKeyForOwnerCall{Call: call}
+}
+
+// MockModelServiceDefaultCloudCredentialKeyForOwnerCall wrap *gomock.Call
+type MockModelServiceDefaultCloudCredentialKeyForOwnerCall struct {
+	*gomock.Call
+}
+
+// Return rewrite *gomock.Call.Return
+func (c *MockModelServiceDefaultCloudCredentialKeyForOwnerCall) Return(arg0 credential.Key, arg1 error) *MockModelServiceDefaultCloudCredentialKeyForOwnerCall {
+	c.Call = c.Call.Return(arg0, arg1)
+	return c
+}
+
+// Do rewrite *gomock.Call.Do
+func (c *MockModelServiceDefaultCloudCredentialKeyForOwnerCall) Do(f func(context.Context, user.Name, string) (credential.Key, error)) *MockModelServiceDefaultCloudCredentialKeyForOwnerCall {
+	c.Call = c.Call.Do(f)
+	return c
+}
+
+// DoAndReturn rewrite *gomock.Call.DoAndReturn
+func (c *MockModelServiceDefaultCloudCredentialKeyForOwnerCall) DoAndReturn(f func(context.Context, user.Name, string) (credential.Key, error)) *MockModelServiceDefaultCloudCredentialKeyForOwnerCall {
+	c.Call = c.Call.DoAndReturn(f)
+	return c
+}
+
 // DefaultModelCloudInfo mocks base method.
 func (m *MockModelService) DefaultModelCloudInfo(arg0 context.Context) (string, string, error) {
 	m.ctrl.T.Helper()

--- a/apiserver/facades/client/modelmanager/services.go
+++ b/apiserver/facades/client/modelmanager/services.go
@@ -8,7 +8,7 @@ import (
 	"time"
 
 	"github.com/juju/juju/apiserver/facade"
-	jujucloud "github.com/juju/juju/cloud"
+	"github.com/juju/juju/cloud"
 	"github.com/juju/juju/core/agentbinary"
 	"github.com/juju/juju/core/assumes"
 	"github.com/juju/juju/core/credential"
@@ -19,7 +19,7 @@ import (
 	"github.com/juju/juju/core/objectstore"
 	"github.com/juju/juju/core/semversion"
 	corestatus "github.com/juju/juju/core/status"
-	coreuser "github.com/juju/juju/core/user"
+	"github.com/juju/juju/core/user"
 	"github.com/juju/juju/core/watcher"
 	"github.com/juju/juju/domain/access"
 	"github.com/juju/juju/domain/blockcommand"
@@ -87,6 +87,11 @@ type ModelService interface {
 	// as that of the model.
 	UpdateCredential(ctx context.Context, uuid coremodel.UUID, key credential.Key) error
 
+	// DefaultCloudCredentialKeyForOwner returns the owner's valid (and non-revoked) credential key for a given cloud
+	// if only one credential exists. If owner has multiple valid cloud credentials a NotFound error is returned as
+	// we cannot determine the default credential.
+	DefaultCloudCredentialKeyForOwner(ctx context.Context, owner user.Name, cloudName string) (credential.Key, error)
+
 	// Model returns the model associated with the provided uuid.
 	Model(ctx context.Context, uuid coremodel.UUID) (coremodel.Model, error)
 
@@ -99,7 +104,7 @@ type ModelService interface {
 	ListAllModels(context.Context) ([]coremodel.Model, error)
 
 	// ListModelsForUser returns a list of models for the given user.
-	ListModelsForUser(context.Context, coreuser.UUID) ([]coremodel.Model, error)
+	ListModelsForUser(context.Context, user.UUID) ([]coremodel.Model, error)
 
 	// ListModelUUIDs returns a list of all model UUIDs in the controller that
 	// are active.
@@ -108,7 +113,7 @@ type ModelService interface {
 	// ListModelUUIDsForUser returns a list of model UUIDs that the supplied
 	// user has access to. If the user supplied does not have access to any
 	// models then an empty slice is returned.
-	ListModelUUIDsForUser(context.Context, coreuser.UUID) ([]coremodel.UUID, error)
+	ListModelUUIDsForUser(context.Context, user.UUID) ([]coremodel.UUID, error)
 
 	// GetModelUsers will retrieve basic information about users with
 	// permissions on the given model UUID.
@@ -116,7 +121,7 @@ type ModelService interface {
 
 	// GetModelUser will retrieve basic information about the specified model
 	// user.
-	GetModelUser(ctx context.Context, modelUUID coremodel.UUID, name coreuser.Name) (coremodel.ModelUserInfo, error)
+	GetModelUser(ctx context.Context, modelUUID coremodel.UUID, name user.Name) (coremodel.ModelUserInfo, error)
 }
 
 // ModelDefaultsService defines a interface for interacting with the model
@@ -175,7 +180,7 @@ type ModelInfoService interface {
 
 	// GetUserModelSummary returns a summary of the current model from the
 	// provided user's perspective.
-	GetUserModelSummary(ctx context.Context, userUUID coreuser.UUID) (coremodel.UserModelSummary, error)
+	GetUserModelSummary(ctx context.Context, userUUID user.UUID) (coremodel.UserModelSummary, error)
 
 	// IsControllerModel returns true if the model is the controller model.
 	IsControllerModel(ctx context.Context) (bool, error)
@@ -189,19 +194,19 @@ type ModelInfoService interface {
 // CredentialService exposes State methods needed by credential manager.
 type CredentialService interface {
 	// CloudCredential returns the cloud credential for the given key.
-	CloudCredential(ctx context.Context, id credential.Key) (jujucloud.Credential, error)
+	CloudCredential(ctx context.Context, id credential.Key) (cloud.Credential, error)
 }
 
 // AccessService defines a interface for interacting the users and permissions
 // of a controller.
 type AccessService interface {
 	// GetUserUUIDByName returns the UUID of the user with the given name.
-	GetUserUUIDByName(context.Context, coreuser.Name) (coreuser.UUID, error)
+	GetUserUUIDByName(context.Context, user.Name) (user.UUID, error)
 	// UpdatePermission updates the access level for a user of the model.
 	UpdatePermission(ctx context.Context, args access.UpdatePermissionArgs) error
 	// LastModelLogin will return the last login time of the specified
 	// user.
-	LastModelLogin(context.Context, coreuser.Name, coremodel.UUID) (time.Time, error)
+	LastModelLogin(context.Context, user.Name, coremodel.UUID) (time.Time, error)
 }
 
 // ModelAgentService provides access to the Juju agent version for the model.

--- a/domain/model/service/package_mock_test.go
+++ b/domain/model/service/package_mock_test.go
@@ -1375,6 +1375,45 @@ func (c *MockStateCreateCall) DoAndReturn(f func(context.Context, model.UUID, mo
 	return c
 }
 
+// DefaultCloudCredentialNameForOwner mocks base method.
+func (m *MockState) DefaultCloudCredentialNameForOwner(arg0 context.Context, arg1 user.Name, arg2 string) (string, error) {
+	m.ctrl.T.Helper()
+	ret := m.ctrl.Call(m, "DefaultCloudCredentialNameForOwner", arg0, arg1, arg2)
+	ret0, _ := ret[0].(string)
+	ret1, _ := ret[1].(error)
+	return ret0, ret1
+}
+
+// DefaultCloudCredentialNameForOwner indicates an expected call of DefaultCloudCredentialNameForOwner.
+func (mr *MockStateMockRecorder) DefaultCloudCredentialNameForOwner(arg0, arg1, arg2 any) *MockStateDefaultCloudCredentialNameForOwnerCall {
+	mr.mock.ctrl.T.Helper()
+	call := mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "DefaultCloudCredentialNameForOwner", reflect.TypeOf((*MockState)(nil).DefaultCloudCredentialNameForOwner), arg0, arg1, arg2)
+	return &MockStateDefaultCloudCredentialNameForOwnerCall{Call: call}
+}
+
+// MockStateDefaultCloudCredentialNameForOwnerCall wrap *gomock.Call
+type MockStateDefaultCloudCredentialNameForOwnerCall struct {
+	*gomock.Call
+}
+
+// Return rewrite *gomock.Call.Return
+func (c *MockStateDefaultCloudCredentialNameForOwnerCall) Return(arg0 string, arg1 error) *MockStateDefaultCloudCredentialNameForOwnerCall {
+	c.Call = c.Call.Return(arg0, arg1)
+	return c
+}
+
+// Do rewrite *gomock.Call.Do
+func (c *MockStateDefaultCloudCredentialNameForOwnerCall) Do(f func(context.Context, user.Name, string) (string, error)) *MockStateDefaultCloudCredentialNameForOwnerCall {
+	c.Call = c.Call.Do(f)
+	return c
+}
+
+// DoAndReturn rewrite *gomock.Call.DoAndReturn
+func (c *MockStateDefaultCloudCredentialNameForOwnerCall) DoAndReturn(f func(context.Context, user.Name, string) (string, error)) *MockStateDefaultCloudCredentialNameForOwnerCall {
+	c.Call = c.Call.DoAndReturn(f)
+	return c
+}
+
 // Delete mocks base method.
 func (m *MockState) Delete(arg0 context.Context, arg1 model.UUID) error {
 	m.ctrl.T.Helper()

--- a/domain/model/service/service.go
+++ b/domain/model/service/service.go
@@ -146,6 +146,11 @@ type State interface {
 	// UpdateCredential updates a model's cloud credential.
 	UpdateCredential(context.Context, coremodel.UUID, credential.Key) error
 
+	// DefaultCloudCredentialNameForOwner returns the owner's default cloud credential name for a given
+	// cloud. If user has multiple (or no) credentials for the specified cloud a NotFound error is returned as
+	// we cannot determine the default credential.
+	DefaultCloudCredentialNameForOwner(ctx context.Context, owner coreuser.Name, cloudName string) (string, error)
+
 	// GetActivatedModelUUIDs returns the subset of model UUIDS from the
 	// supplied list that are activated. If no model uuids are activated then an
 	// empty slice is returned.
@@ -573,6 +578,25 @@ func (s *Service) UpdateCredential(
 	}
 
 	return s.st.UpdateCredential(ctx, uuid, key)
+}
+
+// DefaultCloudCredentialKeyForOwner returns the owner's valid (and non-revoked) credential key for a given cloud
+// if only one credential exists. If owner has multiple valid cloud credentials a NotFound error is returned as
+// we cannot determine the default credential.
+func (s *Service) DefaultCloudCredentialKeyForOwner(ctx context.Context, owner coreuser.Name, cloudName string) (credential.Key, error) {
+	ctx, span := trace.Start(ctx, trace.NameFromFunc())
+	defer span.End()
+
+	name, err := s.st.DefaultCloudCredentialNameForOwner(ctx, owner, cloudName)
+	if err != nil {
+		return credential.Key{}, errors.Capture(err)
+	}
+
+	return credential.Key{
+		Cloud: cloudName,
+		Owner: owner,
+		Name:  name,
+	}, nil
 }
 
 // GetModelLife returns the life associated with the provided uuid.

--- a/domain/model/service/service_test.go
+++ b/domain/model/service/service_test.go
@@ -31,6 +31,7 @@ import (
 	jujuversion "github.com/juju/juju/core/version"
 	"github.com/juju/juju/core/watcher/watchertest"
 	accesserrors "github.com/juju/juju/domain/access/errors"
+	credentialerrors "github.com/juju/juju/domain/credential/errors"
 	domainlife "github.com/juju/juju/domain/life"
 	"github.com/juju/juju/domain/model"
 	modelerrors "github.com/juju/juju/domain/model/errors"
@@ -1224,4 +1225,23 @@ func (s *serviceSuite) TestGetControllerModelUUIDNotFound(c *tc.C) {
 	svc := s.newService(c)
 	_, err := svc.GetControllerModelUUID(c.Context())
 	c.Assert(err, tc.ErrorMatches, "boom")
+}
+
+func (s *serviceSuite) TestDefaultCloudCredentialKeyForOwner(c *tc.C) {
+	defer s.setupMocks(c).Finish()
+
+	svc := s.newService(c)
+
+	s.mockState.EXPECT().DefaultCloudCredentialNameForOwner(gomock.Any(), usertesting.GenNewName(c, "fred"), "cirrus").Return("some-credential", nil)
+
+	fred := usertesting.GenNewName(c, "fred")
+	key, err := svc.DefaultCloudCredentialKeyForOwner(c.Context(), fred, "cirrus")
+	c.Assert(err, tc.ErrorIsNil)
+	c.Assert(key, tc.DeepEquals, credential.Key{Cloud: "cirrus", Owner: fred, Name: "some-credential"})
+
+	s.mockState.EXPECT().DefaultCloudCredentialNameForOwner(gomock.Any(), usertesting.GenNewName(c, "fred"), "cirrus").Return("", credentialerrors.NotFound)
+
+	key, err = svc.DefaultCloudCredentialKeyForOwner(c.Context(), fred, "cirrus")
+	c.Assert(err, tc.ErrorIs, credentialerrors.NotFound)
+	c.Assert(key, tc.DeepEquals, credential.Key{})
 }

--- a/domain/model/service/state_test.go
+++ b/domain/model/service/state_test.go
@@ -18,6 +18,7 @@ import (
 	"github.com/juju/juju/core/version"
 	usererrors "github.com/juju/juju/domain/access/errors"
 	clouderrors "github.com/juju/juju/domain/cloud/errors"
+	credentialerrors "github.com/juju/juju/domain/credential/errors"
 	domainlife "github.com/juju/juju/domain/life"
 	"github.com/juju/juju/domain/model"
 	modelerrors "github.com/juju/juju/domain/model/errors"
@@ -38,6 +39,7 @@ type dummyState struct {
 	users               map[user.UUID]user.Name
 	secretBackends      []string
 	controllerModelUUID coremodel.UUID
+	credentialName      string
 }
 
 func (d *dummyState) CheckModelExists(ctx context.Context, uuid coremodel.UUID) (bool, error) {
@@ -305,6 +307,13 @@ func (d *dummyState) UpdateCredential(
 	}
 
 	return nil
+}
+
+func (d *dummyState) DefaultCloudCredentialNameForOwner(ctx context.Context, owner user.Name, cloudName string) (string, error) {
+	if d.credentialName != "" {
+		return "", credentialerrors.NotFound
+	}
+	return d.credentialName, nil
 }
 
 func (d *dummyState) GetModelUsers(_ context.Context, _ coremodel.UUID) ([]coremodel.ModelUserInfo, error) {

--- a/domain/model/state/controller/state.go
+++ b/domain/model/state/controller/state.go
@@ -22,6 +22,7 @@ import (
 	"github.com/juju/juju/domain"
 	accesserrors "github.com/juju/juju/domain/access/errors"
 	clouderrors "github.com/juju/juju/domain/cloud/errors"
+	credentialerrors "github.com/juju/juju/domain/credential/errors"
 	"github.com/juju/juju/domain/life"
 	"github.com/juju/juju/domain/model"
 	modelerrors "github.com/juju/juju/domain/model/errors"
@@ -2175,4 +2176,57 @@ WHERE uuid = $dbModelUUID.uuid;`,
 	}
 
 	return !invalidCredentialStatus.CredentialInvalid, nil
+}
+
+// DefaultCloudCredentialNameForOwner returns the owner's default cloud credential name for a given
+// cloud. If user has multiple (or no) credentials for the specified cloud a NotFound error is returned as
+// we cannot determine the default credential.
+func (st *State) DefaultCloudCredentialNameForOwner(ctx context.Context, owner user.Name, cloudName string) (string, error) {
+	db, err := st.DB(ctx)
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+
+	var credNames []dbCredKey
+	credQuery := `
+SELECT DISTINCT (cc.name) AS (&dbCredKey.cloud_credential_name)
+FROM   v_cloud_credential cc
+WHERE  cc.owner_name = $ownerAndCloudName.owner_name
+AND    cc.cloud_name = $ownerAndCloudName.cloud_name
+AND    cc.revoked = false
+AND    cc.invalid = false
+`
+	names := ownerAndCloudName{
+		OwnerName: owner.Name(),
+		CloudName: cloudName,
+	}
+	credStmt, err := st.Prepare(
+		credQuery,
+		dbCredKey{},
+		names,
+	)
+	if err != nil {
+		return "", errors.Errorf("preparing select credentials for owner statement: %w", err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+
+		err = tx.Query(ctx, credStmt, names).GetAll(&credNames)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf("loading cloud credentials: %w", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+
+	if len(credNames) == 0 {
+		return "", errors.Errorf("no credentials found: %w", credentialerrors.NotFound)
+	}
+	if len(credNames) > 1 {
+		return "", errors.Errorf("cannot determine default credentials: %w", credentialerrors.NotFound)
+	}
+
+	return credNames[0].CloudCredentialName, nil
 }

--- a/domain/model/state/controller/state_test.go
+++ b/domain/model/state/controller/state_test.go
@@ -29,6 +29,7 @@ import (
 	clouderrors "github.com/juju/juju/domain/cloud/errors"
 	dbcloud "github.com/juju/juju/domain/cloud/state"
 	"github.com/juju/juju/domain/credential"
+	credentialerrors "github.com/juju/juju/domain/credential/errors"
 	credentialstate "github.com/juju/juju/domain/credential/state"
 	"github.com/juju/juju/domain/keymanager"
 	keymanagerstate "github.com/juju/juju/domain/keymanager/state"
@@ -2050,4 +2051,78 @@ func (s *stateSuite) TestHasValidCredentialFalse(c *tc.C) {
 	c.Assert(err, tc.ErrorIsNil)
 	c.Assert(hasValidCredential, tc.IsFalse)
 
+}
+
+func (s *stateSuite) TestDefaultCloudCredentialLabelForOwner(c *tc.C) {
+	st := NewState(s.TxnRunnerFactory())
+
+	accessState := accessstate.NewState(s.TxnRunnerFactory(), loggertesting.WrapCheckLog(c))
+
+	user := usertesting.GenNewName(c, "alice")
+	userUUID := usertesting.GenUserUUID(c)
+	err := accessState.AddUser(
+		c.Context(),
+		userUUID,
+		user,
+		user.Name(),
+		false,
+		userUUID,
+	)
+	c.Check(err, tc.ErrorIsNil)
+
+	credSt := credentialstate.NewState(s.TxnRunnerFactory())
+
+	_, err = st.DefaultCloudCredentialNameForOwner(c.Context(), user, "my-cloud")
+	c.Assert(err, tc.ErrorIs, credentialerrors.NotFound)
+
+	cred := credential.CloudCredentialInfo{
+		Label:    "foobar",
+		AuthType: string(cloud.AccessKeyAuthType),
+		Attributes: map[string]string{
+			"foo": "foo val",
+			"bar": "bar val",
+		},
+	}
+	err = credSt.UpsertCloudCredential(
+		c.Context(), corecredential.Key{
+			Cloud: "my-cloud",
+			Owner: user,
+			Name:  "credential1",
+		},
+		cred,
+	)
+	c.Assert(err, tc.ErrorIsNil)
+
+	name, err := st.DefaultCloudCredentialNameForOwner(c.Context(), user, "my-cloud")
+	c.Assert(err, tc.ErrorIsNil)
+	c.Assert(name, tc.Equals, "credential1")
+
+	err = credSt.UpsertCloudCredential(
+		c.Context(), corecredential.Key{
+			Cloud: "my-cloud",
+			Owner: user,
+			Name:  "credential2",
+		},
+		cred,
+	)
+	c.Assert(err, tc.ErrorIsNil)
+
+	_, err = st.DefaultCloudCredentialNameForOwner(c.Context(), user, "my-cloud")
+	c.Assert(err, tc.ErrorIs, credentialerrors.NotFound)
+
+	cred1UUID, err := credSt.CredentialUUIDForKey(
+		c.Context(),
+		corecredential.Key{
+			Cloud: "my-cloud",
+			Owner: user,
+			Name:  "credential1",
+		},
+	)
+	c.Assert(err, tc.ErrorIsNil)
+	err = credSt.InvalidateCloudCredential(c.Context(), cred1UUID, "test reason")
+	c.Assert(err, tc.ErrorIsNil)
+
+	name, err = st.DefaultCloudCredentialNameForOwner(c.Context(), user, "my-cloud")
+	c.Assert(err, tc.ErrorIsNil)
+	c.Assert(name, tc.Equals, "credential2")
 }

--- a/domain/model/state/controller/types.go
+++ b/domain/model/state/controller/types.go
@@ -347,3 +347,8 @@ type dbModelCredentialInvalidStatus struct {
 type modelUUID struct {
 	UUID coremodel.UUID `db:"uuid"`
 }
+
+type ownerAndCloudName struct {
+	OwnerName string `db:"owner_name"`
+	CloudName string `db:"cloud_name"`
+}

